sw/airborne/firmwares/rotorcraft/navigation.c

```diff
@@ -547,3 +552,107 @@ bool_t nav_set_heading_current(void)
   nav_heading = stateGetNedToBodyEulers_i()->psi;
   return FALSE;
 }
+
+/************** Oval Navigation **********************************************/
+
+/** Navigation along a figure O. One side leg is defined by waypoints [p1] and
+    [p2].
+    The navigation goes through 4 states: OC1 (half circle next to [p1]),
+    OR21 (route [p2] to [p1], OC2 (half circle next to [p2]) and OR12
+    (opposite leg).
+
+    Initial state is the route along the desired segment (OC2).
+*/
+
+enum oval_status oval_status;
+uint8_t nav_oval_count;
+
+void nav_oval_init(void)
+{
+  oval_status = OC2;
+  nav_oval_count = 0;
+}
+
+void nav_oval(uint8_t p1, uint8_t p2, float radius)
+{
+  radius = - radius; /* Historical error ? */
+  int32_t alt = waypoints[p1].enu_i.z;
+  waypoints[p2].enu_i.z = alt;
+
+  float p2_p1_x = waypoints[p1].enu_f.x - waypoints[p2].enu_f.x;
+  float p2_p1_y = waypoints[p1].enu_f.y - waypoints[p2].enu_f.y;
+  float d = sqrtf(p2_p1_x * p2_p1_x + p2_p1_y * p2_p1_y);
+
+  /* Unit vector from p1 to p2 */
+  int32_t u_x = POS_BFP_OF_REAL(p2_p1_x / d);
+  int32_t u_y = POS_BFP_OF_REAL(p2_p1_y / d);
+
+  /* The half circle centers and the other leg */
+  struct EnuCoor_i p1_center = { waypoints[p1].enu_i.x + radius * -u_y,
+           waypoints[p1].enu_i.y + radius * u_x,
+           alt
+  };
+  struct EnuCoor_i p1_out = { waypoints[p1].enu_i.x + 2 * radius * -u_y,
+           waypoints[p1].enu_i.y + 2 * radius * u_x,
+           alt
+  };
+
+  struct EnuCoor_i p2_in = { waypoints[p2].enu_i.x + 2 * radius * -u_y,
+           waypoints[p2].enu_i.y + 2 * radius * u_x,
+           alt
+  };
+  struct EnuCoor_i p2_center = { waypoints[p2].enu_i.x + radius * -u_y,
+           waypoints[p2].enu_i.y + radius * u_x,
+           alt
+  };
+
+  int32_t qdr_out_2 = INT32_ANGLE_PI - int32_atan2_2(u_y, u_x);
+  int32_t qdr_out_1 = qdr_out_2 + INT32_ANGLE_PI;
+  if (radius < 0) {
+    qdr_out_2 += INT32_ANGLE_PI;
+    qdr_out_1 += INT32_ANGLE_PI;
+  }
+  int32_t qdr_anticipation = ANGLE_BFP_OF_REAL(radius > 0 ? -15 : 15);
+
+  switch (oval_status) {
+    case OC1 :
+      nav_circle(&p1_center, POS_BFP_OF_REAL(-radius));
+      if (NavQdrCloseTo(INT32_DEG_OF_RAD(qdr_out_1) - qdr_anticipation)) {
+        oval_status = OR12;
+        InitStage();
+//        LINE_START_FUNCTION;
```

Comment:
I think it would be worth keeping those (for the cam modules).
You can let them default to something empty like in fixedwing/navigation.c